Created MainScraper

btkcodedev opened the pull request:
### Backend Scraper for flipkart
_Libraries of Beautifulsoup and requests are used for scraping the contents_ 

- The scraper make use of the Flipkart site in which 2 types of styles used in the sections are scraped

- As an example, Laptops, mobiles, Cameras that uses the style 1 and headphones, Air conditioners uses the style 2 are scraped,
Both the styles can scrap several other categories.

> Logic
> The styles in the flipkart mainly divided into two and both are scraped on the basis of the tag name and pattern
> If the first style doesn't match the pattern, then no result will be forwarded to the list, thus the second styling is checked

- The number of pages to scrap can be changed to the limit in line 10
`for i in range(1,5):  #use n+1 pages for clearing index out of range error`

- The number of pages can be set in the line 24 as for loop range
`for i in range(1,limit):  #only use n-1 pages which are proxy safe and error free in the list`
  `myurl= urlss[i]`

- The results will be the Link,Title,Rating,Amount that will be listed in different python lists in strict order along with number of things scraped at last. (Also the csv and io are imported which can be used when file management is necessary)
`print(l) #Links`
`print(title) #Title`
`print(rat) #Rating`
`print(amount) #Amount`
`print(len(amount))  #Number of items scraped`
`print("pages")`

- The frontend can make use of the variables in the URL in line 16 for changing the need for the user along some dropdown boxes 
`urlss=urlss+[cams]  # Change the variable with a dropdown box when merging with its front end`

-Hope it gives an addition for the project.

btkcodedev committed:
Created mainscaper.py

The heart of this scraper mainly focuses on flipkart
Generalization of representational styles in flipkart sections and divided into two ways with example 5 urls
First 3 url's uses first style and next 2 uses second style in their representation
Finally
Link,Title,Rating,Amount are listed in different lists in strict order along with number of things scraped at last

btkcodedev committed:
Merge pull request #1 from btkcodedev/btkcodedev-scaper-1

Created mainscaper.py

anandbaburajan commented:
@btkcodedev Hey! Thanks for the contribution! The comments helped a lot. I've a few suggestions, since the project maintainer and other members don't seem to be active:

You can refactor `scaper.py` so that it can be used as the back-end of this project . You can do that by modularizing the existing code into their own functions. Example: `get_urls()`, `scrape_page()` etc. and replacing the `print` statements with `return`s. The functions can take parameters like number of pages, product category, etc. This way, your functions can be easily used by the front-end devs.  :) You can also rename the file to `flipkart_scraper.py` right? Also, it's better if you put the comments above the statements, instead of in-line. Also, no need for comments for the `import` statements.

Also, next time onwards, please make sure to create a new branch for your PRs.

Thanks again!

btkcodedev committed:
Changed as functions

btkcodedev committed:
Merge pull request #2 from btkcodedev/btkcodedev-patch-1

Changed as functions

btkcodedev commented:
@AnandBaburajan 
The changes are made to the code
Also it can be interpreted and results can be examined by the front end developers

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+def get_url(dropdownselection,limit):
+  urlss=[] 
+  if dropdownselection==1:
+    for i in range(1,limit):  #use n+1 pages for clearing index out of range error
```
Put comment on top of the statement here and similarly for all other statements.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+  urlss=[] 
+  if dropdownselection==1:
+    for i in range(1,limit):  #use n+1 pages for clearing index out of range error
+      laps='https://www.flipkart.com/laptops/pr?sid=6bo,b5g&otracker=categorytree&page='+str(i)
+      urlss=urlss+[laps]
+  elif dropdownselection==2:
+    for i in range(1,limit): 
+      mobs='https://www.flipkart.com/mobiles/pr?sid=tyy%2C4io&otracker=categorytree&page='+str(i)
+      urlss=urlss+[mobs]
+  elif dropdownselection==3:
+     for i in range(1,limit): 
+      cams='https://www.flipkart.com/cameras/pr?sid=jek,p31&otracker=categorytree&page='+str(i)
+      urlss=urlss+[cams]
+  elif dropdownselection==4:
+     for i in range(1,limit): 
+      hpo='https://www.flipkart.com/headphones/pr?sid=fcn&otracker=categorytree&page='+str(i)
+      urlss=urlss+[hpo]
+  elif dropdownselection==5:
+     for i in range(1,limit): 
+      acs='https://www.flipkart.com/air-conditioners/pr?sid=c54&otracker=categorytree&page='+str(i)
+      urlss=urlss+[acs]
```
Creating a dictionary which stores all those links and then accessing the values would look better. Also, since `for i in range(1,limit)` is repeated a lot of times here. Maybe you can refactor the code so that it's used just once. (You can put the conditions inside the loop). Also, you can use `switch - case` here with a `default`. Would be more readable.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+# The front end can make use of the passing parameters
+# In the dropdown list, the developer can give the Words like Laptops, Mobiles, Cameras, Headphones, Air conditioners etc
+# Also the user can define the limit of number of pages to scrap [It is used in both functions]
+def get_url(dropdownselection,limit):
```
It'd be better if the Python variables were in camel_case. Like, `dropdown_selection`

anandbaburajan replied:
Also, it's `get_urls` right?

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def scrape_page(urlss,limit-1):
+  import requests     #Helps to send HTTP request(connections)
+  from urllib.request import urlopen as ureq #Helps to get connected with Specified URL's
+  from bs4 import BeautifulSoup as soup #Helps to extract webpages using HTML Parser function
+  import smtplib #Helps to send out mails using SMTP
+  import io  #Helps tp encode the characters into specified format
+  import csv  #Helps to read csv file
```
A good practice is top have all of your import statements at the top of your file, instead of inside functions. You can also remove the comments because interested people can search for that on their own. (better looking)

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+          print("\n")
+          
+          
+  if title==[]:  #for different type of page
```
The best Pythonic way of checking empty list: `if not title:`

anandbaburajan replied:
Also, can you look into refactoring the part hereafter into its own function? Would help in reducing the code complexities (you can check your PRs codeclimate results)

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+  import io  #Helps tp encode the characters into specified format
+  import csv  #Helps to read csv file
+  #opens up the web connection with specified URL
+  title=[]
+  amount=[]
+  l=[]
+  spec=[]
+  rat=[]
```
Since these are lists, please give more meaningful names likes `titles`, `amounts` and please give better names to the rest

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+  return (urlss)
+
+
+def scrape_page(urlss,limit-1):
```
The `-1` looks a bit weird, can you refactor the code to deal with this?

anandbaburajan replied:
Also can you add a docstring which helps the user to understand what this function takes and returns? Thanks!

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+            rat=rat+["No rating"]
+          specs=headdata.findAll("li",{"class":"tVe95H"})
+          for spec in specs:
+            print(spec.text)
+          print("\n")
```
Do we need these `prints`?

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,120 @@
+      page_soup = soup(page_html,"html.parser")
+      containers = page_soup.findAll("div",{"class":"bhgxx2 col-12-12"})
+      for container in containers:
+          
+        datas=container.findAll("div",{"class":"_3liAhj"})
+
```
Please use a formatter for Python to take care of consisten spacing between lines and variables, etc.

anandbaburajan reviewed (changes requested):
Thanks! Great work~ Looking forward to see this integrated with the client side!

btkcodedev committed:
Update functions for shortening the code

Shortened the code
Used dictionary for getting url's
Decreased the time complexity
Removed unnecessary print statements
Removed unnecessary comments
Refactored code to fix the limit variable
Refactored function names and passing parameters

btkcodedev committed:
Merge pull request #3 from btkcodedev/btkcodedev-patch-2

Update functions for shortening the code

btkcodedev commented:
@AnandBaburajan The required changes are done
Now the whole code can be called with one function
Eg; scrape_page(get_urls(4,6),6)
Thus it will be easier for the front end to merge
More details are shared in the latest commit

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,113 @@
+            ratings=ratings+["No rating"]
+          
+  #for different type of page  
+  if not title:  
```
Did you look into making this into its own function? Would help with decreasing the complexity (check the codeclimate results). I think `scrape_page` can be split into two-three function! Thanks!

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,113 @@
+  # Change the variable with a dropdown box when merging with its front end according to the need
+  return (urlss)
+
+def scrape_page(urlss,limit):
```
Like I mentioned earlier, can you please add a docstring for the functions? https://www.geeksforgeeks.org/python-docstrings/

anandbaburajan reviewed (changes requested):
@btkcodedev Thanks! Rest of the changes look good! I've added a few comments then this PR can be merged!

btkcodedev committed:
Update flipkart_scraper.py

Separated scrape_page() into two functions
Added Docstring

btkcodedev committed:
Merge pull request #4 from btkcodedev/btkcodedev-patch-3

Update flipkart_scraper.py

btkcodedev commented:
@AnandBaburajan The required changes are done,
- Complexity reduced:
The scrape_page() was split into two and complexity reduced from 46 to 14 and 24 for two functions.
But that will be not a problem, I think because the second type function will only execute if the first type fails
And it totally depends on the number of pages scraped, and that will be not more than 10 I guess.
- Docstrings added:
Also now the frontend developers can easily identify the flow using the docstrings and comments

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+# Also the user can define the limit of number of pages to scrap [It is used in both functions]
+
+def get_urls(DropDownSelection,limit):
+  '''The First function executed when scrape_page(get_urls(dds,limit),limit) is called
+  is the get_url, it makes a list of links according to the limit and selection
+  in order to forward to the scraper part'''
```
The best way to use doctrings is the following format:

```python
def my_function(arg1): 
    """ 
    Summary line. 
  
    Extended description of function. 
  
    Parameters: 
    arg1 (int): Description of arg1 
  
    Returns: 
    int: Description of return value 
  
    """
  
    return arg1 
```

Also, "unlike conventional source code comments, the docstring should describe what the function does, not how". So here, it would be better if you describe in simple terms what `get_urls` takes as input and what it returns. Rest of the "good to know" things can be put in a comment somewhere else.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+  return (urlss)
+
+def type2(urlss,limit):
+  '''When the type1 script fails to identify the sections, the type2 will be executed, most 
+  filpkart section relies on these two types, thus the scraper follows same parameters and gives 
+  the output'''
```
Similarly fix the docstring here.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+  # Change the variable with a dropdown box when merging with its front end according to the need
+  return (urlss)
+
+def type2(urlss,limit):
```
Can you change it to `scrape_page_type_2`? Would be more readable I think.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+        amount=amount+[realprice]
+  return (linkss, title, ratings, amount, len(amount)) 
+
+def scrape_page(urlss,limit):
```
Can you change it to `scrape_page_type_1`? Would be more readable I think.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+  return (linkss, title, ratings, amount, len(amount)) 
+
+def scrape_page(urlss,limit):
+  '''The get_urls return the url's list to the scrape_page() function
+  This function contains the type1 script of the scraper, which is one of the 
+  pattern in which the details of the products are arranged'''
```
Similarly fix the docstring here.

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+   answer = type2(urlss,limit)
+   return answer
+  else:
+   '''And finally the function returns the list of links,ratings,amount,number of products.'''
```
This can be a simple comment because docstrings are only used to describe functions, classes, etc :)

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,133 @@
+  #Returns the Links, Ratings, Amount, Number of products scraped in list with strict order
+   return (linkss, title, ratings, amount, len(amount)) 
+
+# Call the function like : scrape_page(get_urls(4,6),6), where 4 specifies the dropdownselection and 6 is the number of pages
```
Can you put it somwehere at the top?

anandbaburajan reviewed (changes requested):
@btkcodedev I've added some final minor comments after which this would be good to merge! Thanks!

btkcodedev committed:
Update flipkart_scraper.py

Cleansed the documentation

btkcodedev committed:
Merge pull request #5 from btkcodedev/btkcodedev-patch-4

Update flipkart_scraper.py

btkcodedev commented:
@AnandBaburajan Required changes are done

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,140 @@
+# Also the user can define the limit of number of pages to scrap [It is used in both functions]
+
+def get_urls(DropDownSelection,limit):
+  '''When get_url() is called, it makes a list of links according to the limit and selection
+  in order to forward to the scraper part
+  Argument1= DropDownSelection :: Selects the category from the frontend
+  Argument2= limit :: The limit for the number of web pages selected by user or by developer in frontend'''
```
Please put the `"""` on the top and below the description. Also, I see you provided a docstring for the return statement below which is incorrect. I suggest to go with something like:

```python
def get_urls(DropDownSelection, limit):
    """ 
    Function to make a list of links according to the limit and selection.
  
    Parameters: 
    DropDownSelection (int): Selects the category from the frontend
    limit (int): The limit for the number of web pages selected by user or by developer in frontend
  
    Returns: 
    urlss (List): List names urlss that contains the list of links based on the given category.
    """
  
    return arg1 
```

anandbaburajan replied:
@btkcodedev please correct the docstring as suggested in the example

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,140 @@
+  return (urlss)
+
+def scrape_page_type_2(urlss,limit):
+  '''When the type1 script fails to identify the sections, the type2 will be executed, most 
+  filpkart section relies on these two types, thus the scraper follows same parameters and gives 
+  the output
+  Argument1= urlss :: The list of links to be scraped that was the result from get_urls
+  Argument2= limit :: The limit for the number of web pages selected by user or by developer in frontend'''
```
Fix docstring here

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,140 @@
+    item = switch[DropDownSelection] + str(i)
+    urlss=urlss+[item]
+  # Change the variable with a dropdown box when merging with its front end according to the need
+  '''return(list)::
+  Returns the list names urlss that contains the list of links based on the given category'''
```
Docstrings are only used to describe classes, functions etc. Please remove this one.

anandbaburajan replied:
Docstrings aren't used here, please remove this and refer to the article I shared earlier

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,140 @@
+        realprice=nextcontainers[0].text
+        #Found price
+        amount=amount+[realprice]
+  '''return(list)::
+  Returns the list Links, Ratings, Amount, Number of products scraped in list with strict order'''
+  return (linkss, title, ratings, amount, len(amount)) 
```
Fix docstring as suggested

anandbaburajan commented on flipkart_scraper.py:
```diff
@@ -0,0 +1,140 @@
+  Returns the list Links, Ratings, Amount, Number of products scraped in list with strict order'''
+  return (linkss, title, ratings, amount, len(amount)) 
+
+def scrape_page_type_1(urlss,limit):
+  '''The get_urls return the url's list to the scrape_page() function
+  This function contains the type1 script of the scraper, which is one of the 
+  pattern in which the details of the products are arranged
+  Argument1= urlss :: The list of links to be scraped that was the result from get_urls
+  Argument2= limit :: The limit for the number of web pages selected by user or by developer in frontend'''
```
Fix docstring as suggested

anandbaburajan reviewed (changes requested):
@btkcodedev I've added some comments. Please go through https://www.geeksforgeeks.org/python-docstrings/ for more info on docstrings.

btkcodedev committed:
Update flipkart_scraper.py

Refactored the tags due to updations in the flipkart site
Added docstrings

btkcodedev committed:
Merge pull request #7 from btkcodedev/btkcodedev-patch-6

Update flipkart_scraper.py

btkcodedev committed:
Merge branch 'FOSS-Cell-GECPKD:master' into master
